src/TAMSEquationSystem.C

```diff
@@ -0,0 +1,615 @@
+void
+TAMSEquationSystem::reinitialize_linear_system()
+{
+  // Nothing to do here...
+}
+
+//--------------------------------------------------------------------------
+//-------- solve_and_update ------------------------------------------------
+//--------------------------------------------------------------------------
+void
+TAMSEquationSystem::solve_and_update()
+{
+  // Nothing to do here...
+}
+
+//--------------------------------------------------------------------------
+//-------- initial_work ----------------------------------------------------
+//--------------------------------------------------------------------------
+void
+TAMSEquationSystem::initial_work()
+{
+  compute_metric_tensor();
+
+  stk::mesh::MetaData& meta_data = realm_.meta_data();
+
+  // Initialize average_velocity, avg_dudx and avg_Prod
+  // We don't want to do this on restart where TAMS fields are present
+  if (resetAverages_) {
+    const int nDim = meta_data.spatial_dimension();
+
+    // Copy velocity to average velocity
+    VectorFieldType &avgU = avgVelocity_->field_of_state(stk::mesh::StateNP1);
+    const auto& U = *meta_data.get_field<VectorFieldType>(stk::topology::NODE_RANK, "velocity");
+    field_copy(realm_.meta_data(), realm_.bulk_data(), U, avgU, realm_.get_activate_aura());
+
+    // Copy dudx to average dudx
+    GenericFieldType &avgDudx = avgDudx_->field_of_state(stk::mesh::StateNP1);
+    const auto& dudx = *meta_data.get_field<GenericFieldType>(stk::topology::NODE_RANK, "dudx");
+    field_copy(realm_.meta_data(), realm_.bulk_data(), dudx, avgDudx, realm_.get_activate_aura());
+
+    ScalarFieldType* turbKinEne_ = meta_data.get_field<ScalarFieldType>(
+      stk::topology::NODE_RANK, "turbulent_ke");
+    ScalarFieldType* tvisc_ = meta_data.get_field<ScalarFieldType>(
+      stk::topology::NODE_RANK, "turbulent_viscosity");
+    ScalarFieldType& tkeNp1 = turbKinEne_->field_of_state(stk::mesh::StateNP1);
+
+    // define some common selectors
+    stk::mesh::Selector s_all_nodes =
+      (meta_data.locally_owned_part() | meta_data.globally_shared_part()) &
+      stk::mesh::selectField(*avgDudx_);
+
+    stk::mesh::BucketVector const& buckets =
+      realm_.get_buckets(stk::topology::NODE_RANK, s_all_nodes);
+    for (stk::mesh::BucketVector::const_iterator ib = buckets.begin();
+         ib != buckets.end(); ++ib) {
+      stk::mesh::Bucket& b = **ib;
+      const stk::mesh::Bucket::size_type length = b.size();
+
+      double* tke = stk::mesh::field_data(tkeNp1, b);
+      double* tvisc = stk::mesh::field_data(*tvisc_, b);
+      double* avgProd = stk::mesh::field_data(*avgProduction_, b);
+      double* rho = stk::mesh::field_data(*avgDensity_, b);
+
+      for (stk::mesh::Bucket::size_type k = 0; k < length; ++k) {
+        // Initialize average production to mean production
+        const double* avgDudx = stk::mesh::field_data(*avgDudx_, b[k]);
+        double* tij = new double[nDim * nDim];
```

Comment:
Can we use constant size arrays or `std::vector` here instead of new delete

Reply:
done.